### ISSUE TITLE
Classifier feature engineering/evaluation

### ISSUE THREAD

**Issue by wbwakeman** [AllenInstitute/croissant#40]
How far can we get with logistic regression improvements in 1 day?

Tasks:
- [x] train classifier with current coded features
- [x] generate presentation output
- [x] Investigate feature improvements (time-boxed to maximum of 2 developer days)

Acceptance criteria:
- [x] Presentation output of current features
- [x] branch with new features from this effort
- [x] includes description of what has been added and why
- [x] Notebooks used for this should be saved to persistent storage somewhere

**Comment by kschelonka** [comment]
The full notebook is uploaded to s3. Includes some data exploration, hyperparameter tuning using `hyperopt` (tested random forest  and xgboost), plots, feature engineering examples

`s3://prod.slapp.alleninstitute.org/2-line 2-project/feature_engineering/data_exploration_feature_eng_2020-07-20-11-08.ipynb`

**Summary:**
* Identified 13 features related to traces and mask morphology
* Extracted spikes from calcium traces using https://github.com/j-friedrich/OASIS (which is incorporated into the CaImAn library)
* Random forest, 5-fold cross-validation optimizing for average precision (proxy for area under the PR curve) using hyperopt. 
* Some of the new features generated had high importances in the random forest model, indicating their usefulness
* Was able to get a pretty decent model, but there is still room for improvement
* Should re-engage scientists to identify most important metrics for optimization
* Should use a larger feature space with more iterations for hyperparameter tuning

**Generated features**: 
* avg_adjacent_n: the average number of pixels adjacent to each pixel in the mask (includes diagonals, maximum=8) -- proxy for the "filled-in-ness" of the mask pixels
* std_adjacent_n: standard deviation of the distribution of pixels adjacent to each pixel in the mask (as described above)
* skew_adjacent_n: skew of the distribution of pixels adjacent to each pixel in the mask (as described above)
* last_tenth_trace_skew: the skew of the last tenth of the trace. This roughly approximates when the "fingerprint stimuli" are presented, where more of the cells are likely to be activated
* avg_spike_amp: average of the mean-subtracted spike amplitudes (extracted using OASIS)
* std_spike_amp: the standard deviation of the mean-subtracted spike amplitudes (extracted using OASIS)
* avg_spike_len: average number of frames each spike lasted (extracted using OASIS)
* spike_count: total number of extracted spikes (using OASIS)
* roi_data_std: standard deviation of the pixels in the weighted mask (values from Suite2p)
* roi_data_skew: skew of the pixels in the weighted mask (values from Suite2p)
* avg_nearest_dist: average Euclidean distance of each pixel to its nearest neighbor for a mask (rough approximation of mask pixel "density")
* std_nearest_dist: standard deviation of Euclidean distance of each pixel to its nearest neighbor for a mask (rough approximation of mask pixel "density")
* skew_nearest_dist: skew of the Euclidean distance of each pixel to its nearest neighbor for a mask (rough approximation of mask pixel "density")

**Best parameter set:**

`{'bootstrap': True, 'max_depth': 40, 'max_features': 0.5, 'min_samples_leaf': 10, 'min_samples_split': 20, 'n_estimators': 25}`

**Train summary:**
```
Test Accuracy: 0.8368421052631579
Test recall: 0.7692307692307693
Test avg precision: 0.7836032388663968
Test AUC: 0.8341103341103342
Train Accuracy: 0.8941798941798942
Train recall: 0.8456591639871383
Train avg precision: 0.8174187079958511
Train AUC: 0.8868745258137938
```

**Model feature importances:**
```
[('avg_adjacent_n', 0.3270389720830048),
 ('skew_adjacent_n', 0.17242285988311368),
 ('roi_ellipticalness', 0.14234041896469612),
 ('last_tenth_trace_skew', 0.08687341136672706),
 ('std_spike_amp', 0.03824486539586723),
 ('avg_spike_len', 0.02878557129790896),
 ('full_genotype', 0.025752729812740856),
 ('depth', 0.025419801072027133),
 ('std_adjacent_n', 0.023772342799012396),
 ('roi_data_std', 0.02299103513224442),
 ('avg_spike_amp', 0.021437207422193627),
 ('roi_data_skew', 0.016952828654097057),
 ('roi_len', 0.014070367636467808),
 ('trace_skew', 0.013923947011934844),
 ('std_nearest_dist', 0.010693745945718052),
 ('spike_count', 0.008667749900153518),
 ('roi_height', 0.008148682825574266),
 ('targeted_structure', 0.006168795488666053),
 ('skew_nearest_dist', 0.005303341929219612),
 ('roi_area', 0.0009913253786326115)]
```

**Confusion matrix (normalized ratios):**
![image.png](https://images.zenhubusercontent.com/5d8d3df983bdab000177f354/77635492-3434-42c5-9e6a-19366a96d20b)

**Comment by kschelonka** [comment]
Note on this criteria:
 branch with new features from this effort

I am going to call this finished because there are example implementations in the notebook. I think the finalized implementation should be a separate story that is reviewed and merged with our usual processes.
